File: DSA/Array/0121_Best_Time_to_Buy_and_Sell_Stock/cpp/solution.cpp
// 0121. Best Time to Buy and Sell Stock
// Difficulty : Easy
// Tags       : array, dynamic-programming
// Solved on  : 2019-07-23
// Attempt    : #1
// LeetCode   : https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution {
public:
    
    int maxProfit(vector<int>& prices) {
        int mn=INT_MAX;
        int mx=0;
        for(int i=0;i<prices.size();i++){
            if(prices[i]<mn){
                mn=prices[i];
               
            }
             mx=max(prices[i]-mn,mx);
           
        }
        return mx;
    }
    
};

File: DSA/Array/0121_Best_Time_to_Buy_and_Sell_Stock/solution.php
// 0121. Best Time to Buy and Sell Stock
// Difficulty : Easy
// Tags       : array, dynamic-programming
// Solved on  : 2026-05-06
// LeetCode   : https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution {

    /**
     * @param Integer[] $prices
     * @return Integer
     🔷 Step-by-Step Example
        prices = [7, 1, 5, 3, 6, 4]
        | Day | Price | min_price_so_far | Profit if sold today | max_profit |
        | --- | ----- | ---------------- | -------------------- | ---------- |
        | 0   | 7     | 7                | 0                    | 0          |
        | 1   | 1     | 1                | 0                    | 0          |
        | 2   | 5     | 1                | 4                    | 4          |
        | 3   | 3     | 1                | 2                    | 4          |
        | 4   | 6     | 1                | 5                    | 5          |
        | 5   | 4     | 1                | 3                    | 5          |
    */
    function maxProfit($prices) {
        $mn = PHP_INT_MAX;
        $mx = PHP_INT_MIN;
        foreach($prices as $p){
            $mn = min($mn,$p);
            $mx = max($p-$mn,$mx);
        }
        return $mx;
    }
}